### ISSUE TITLE
BigGAN minibatch sizes: gradient accumulation support?

### ISSUE THREAD

**Issue by gwern** [AaronLeong/BigGAN-pytorch#1]
In the [BigGAN paper](https://arxiv.org/pdf/1809.11096.pdf), one of the important features is the use of large minibatches, _n_=2048 for most of the results. Table 1 shows that FID/Inception improve considerably with, among other things, minibatches going from 256 to 2048 (FID halves, IS almost doubles). In comparison, the other tweaks appear to have smaller benefits.

Obviously, most people don't have a 512-core TPUv3 slice to train on and can't train those minibatches directly. (I'm getting minibatches of 33 on my 2 GPUs right now.) But accumulating gradients across multiple minibatches should be near-identical (at a wallclock multiple slowdown), and might be important for getting the best results. So gradient accumulation should be supported.

As I understand it in PyTorch, gradient accumulation is the default behavior of `.backward()`, and all that is necessary is to add an inner loop to `train.py`'s `train` loop around the D/G updates, where `d/g_optimizer.step()` gets run every _i_ iterations, where _i_ is the total minibatch / actual minibatch. So something like 'for step ... / for i ... / ... / d/g_optimize.step()'.

**Comment by AaronLeong** [comment]
Thanks for your issue,  This project is not intended to reproduce the results of the paper, my goal is only to achieve some of the realizable tricks in the paper.

**Comment by jcpeterson** [comment]
@gwern why only 33? why not the default 64?

**Comment by gwern** [comment]
> This project is not intended to reproduce the results of the paper, my goal is only to achieve some of the realizable tricks in the paper.

I see. But the change is simple and may be important for better performance. It's worth supporting.

> why not the default 64?

33 is what seems to fit on mine; had to change it because 64 did not fit.

Why? AaronLeong may be using 4 GPUs, given that one of the README examples specifies `--gpus 0,1,2,3`, in which case it's no surprise he can run twice the minibatch size I can on 2 GPUs. Another, probably less likely, possibility is the CuDNN use: `cudnn.benchmark = True` in `main.py`, and IIRC, that causes a space-time tradeoff (I vaguely recall another GAN I was using ultimately chose to specifically disable CuDNN to allow larger but somewhat slower minibatches). Perhaps his CuDNN install is broken or tweaked somehow, so it has lower memory usage.
